File: DepthFirstSearch/329_LongestIncreasingPathinaMatrix.java
public class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int m = matrix.length;
        if (m == 0) return 0;
        int n = matrix[0].length;
        int[][] cache = new int[m][n];
        int longest = 1;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                longest = Math.max(longest, getLongest(matrix, i, j, Integer.MIN_VALUE, cache));
            }
        }
        return longest;
    }

    public int getLongest(int[][] matrix, int i, int j, int pre, int[][] cache) {
        if (i < 0 || i >= matrix.length || j < 0 || j >= matrix[0].length || matrix[i][j] <= pre) return 0;
        if (cache[i][j] != 0) return cache[i][j];
        int longest = 1;
        longest = Math.max(longest, 1 + getLongest(matrix, i - 1, j, matrix[i][j], cache));
        longest = Math.max(longest, 1 + getLongest(matrix, i + 1, j, matrix[i][j], cache));
        longest = Math.max(longest, 1 + getLongest(matrix, i, j - 1, matrix[i][j], cache));
        longest = Math.max(longest, 1 + getLongest(matrix, i, j + 1, matrix[i][j], cache));
        cache[i][j] = longest;
        return longest;
    }
}

File: DepthFirstSearch/329_LongestIncreasingPathinaMatrix.py
class Solution(object):

    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if len(matrix) == 0:
            return 0
        longest = 1
        dic = {}
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                longest = max(longest, self.getLongest(matrix, i, j, -2147483648, dic))
        # test = self.getLongest(matrix, 0, 0, -2147483648, dic)
        return longest


    def getLongest(self, matrix, i, j, curr, dic):
        if i < 0 or i >= len(matrix) or j < 0 or j >= len(matrix[0]):
            # print "ii"
            return 0
        if matrix[i][j] <= curr:
            return 0
        key = str(i*len(matrix[0])+j)
        if dic.has_key(key):
            return dic.get(key)
        top = 1 + self.getLongest(matrix, i - 1, j, matrix[i][j],dic)
        left = 1 + self.getLongest(matrix, i, j-1, matrix[i][j],dic)
        bottom = 1 + self.getLongest(matrix, i+1, j, matrix[i][j],dic)
        right = 1 + self.getLongest(matrix, i, j+1, matrix[i][j],dic)
        # print top,left,bottom,right
        longest = max(top, left, bottom, right)

        dic[key] = longest
        return longest